src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -1583,31 +1581,22 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
     }
 
     /** Compile list of abstract files. */
-    def compileFiles(files: List[AbstractFile]): Unit = {
-      try {
-        val snap = profiler.beforePhase(Global.InitPhase)
-        val sources = files map getSourceFile
-        profiler.afterPhase(Global.InitPhase, snap)
-        compileSources(sources)
-      }
-      catch {
-        case ex: InterruptedException => reporter.cancelled = true
-        case ex: IOException => globalError(ex.getMessage())
-      }
-    }
+    def compileFiles(files: List[AbstractFile]): Unit = compileFilesInternal(files.map(getSourceFile))
 
     /** Compile list of files given by their names */
-    def compile(filenames: List[String]): Unit = {
+    def compile(filenames: List[String]): Unit =
+      compileFilesInternal {
+        if (settings.script.isSetByUser && filenames.size > 1) {
+          globalError("can only compile one script at a time")
+          Nil
+        }
+        else filenames.map(getSourceFile)
+      }
+
+    private def compileFilesInternal(getSources: => List[SourceFile]): Unit = {
```

Comment:
OK, now I see that you want to handle exceptions on getting sources and compiling sources.

For me, it would be simpler to see it, if it were `compileFilesInternal(safely(getMySources))` and then here `safely { compileSources(sources) }`. Where `safely` de-dups the try-catch. (Naming it left as an exercise.)

Reply:
I was just looking at `Global.InitPhase` and noticed it's in two quasi identical blocks.

Could do:

```scala
    private val safely: Exception.Catch[Unit] = {
      val cancellingCatch = Exception
          .catchingPromiscuously(classOf[InterruptedException])
          .withApply(_ => reporter.cancelled = true)
      val ioLogCatch = Exception.handling(classOf[IOException]).by(ex => globalError(ex.getMessage()))
      cancellingCatch.or(ioLogCatch)
    }

    def profilePhase[A](ph: Phase)(x: => A): A = {
      val snap = profiler.beforePhase(ph)
      try x finally profiler.afterPhase(ph, snap)
    }

    /** Compile list of abstract files. */
    def compileFiles(files: List[AbstractFile]): Unit = safely {
      val sources = profilePhase(Global.InitPhase)(files.map(getSourceFile))
      compileSources(sources)
    }

    /** Compile list of files given by their names */
    def compile(filenames: List[String]): Unit = safely {
      val sources = profilePhase(Global.InitPhase) {
        if (settings.script.isSetByUser && filenames.size > 1) {
          globalError("can only compile one script at a time")
          Nil
        }
        else filenames.map(getSourceFile)
      }
      compileSources(sources)
    }
```

... but I'm not if that's preferred... 👨‍🎨